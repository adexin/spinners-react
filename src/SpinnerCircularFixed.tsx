import React, { CSSProperties } from 'react';
import { secondaryColorDefaultProps, SecondaryColorSpinnerProps } from './helpers';
import { SpinnersProps, withSharedProps } from './withSharedProps';

import './SpinnerCircularFixed.css';

export type SpinnerCircularFixedProps = SpinnersProps & SecondaryColorSpinnerProps;

const Component = ({
  secondaryColor = secondaryColorDefaultProps.secondaryColor,
  speed = secondaryColorDefaultProps.speed,
  still = secondaryColorDefaultProps.still,
  thickness = secondaryColorDefaultProps.thickness,
  ...svgProps
}: SecondaryColorSpinnerProps) => {
  const strokeWidth = 4 * (thickness / 100);
  const circleStyle: CSSProperties = !still
    ? { animation: `spinners-react-circular-fixed ${140 / speed}s linear infinite` }
    : {};

  return (
    <svg fill="none" {...svgProps} viewBox="0 0 66 66">
      <circle
        cx="33"
        cy="33"
        fill="none"
        r="28"
        stroke={secondaryColor}
        strokeWidth={strokeWidth}
      />
      <circle
        cx="33"
        cy="33"
        fill="none"
        r="28"
        stroke="currentColor"
        strokeDasharray="40, 134"
        strokeDashoffset="325"
        strokeLinecap="round"
        strokeWidth={strokeWidth}
        style={circleStyle}
      />
    </svg>
  );
};

export const SpinnerCircularFixed = withSharedProps(Component);
