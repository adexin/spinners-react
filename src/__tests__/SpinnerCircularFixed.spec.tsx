import { render } from '@testing-library/react';
import React from 'react';

import { SpinnerCircularFixed } from '../SpinnerCircularFixed';

describe('SpinnerCircularFixed', () => {
  const color = 'red';

  it('matches the snapshot', () => {
    const { container } = render(
      <SpinnerCircularFixed color="#fff" size={100} speed={10} thickness={50} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot with default props', () => {
    const { container } = render(<SpinnerCircularFixed />);

    expect(container).toMatchSnapshot();
  });

  it('renders null if disabled', () => {
    const { container } = render(<SpinnerCircularFixed enabled={false} />);

    expect(container.firstChild).toBe(null);
  });

  it('renders still if specified', () => {
    const { container } = render(<SpinnerCircularFixed still />);
    const circles = container.querySelectorAll('circle');

    expect(circles[1].style.animation).toBe('');
  });

  it('passes props to styles', () => {
    const size = '20%';
    const thickness = 40;
    const speed = 50;
    const { container } = render(
      <SpinnerCircularFixed
        color={color}
        secondaryColor="blue"
        size={size}
        speed={speed}
        thickness={thickness}
      />,
    );
    const circles = container.querySelectorAll('circle');

    expect(container.firstChild).toHaveStyle({ width: size });
    expect(circles[0]).toHaveAttribute(
      'stroke-width',
      `${4 * (thickness / 100)}`,
    );
    expect(circles[1].style.animation.includes(`${140 / speed}`)).toBe(true);
  });

  it('passes svg props overriding styles', () => {
    const className = 'test-class';
    const { container } = render(
      <SpinnerCircularFixed
        className={className}
        color="green"
        style={{ color }}
      />,
    );

    expect(container.firstChild).toHaveClass(className);
    expect(container.firstChild).toHaveStyle({ color });
  });
});
