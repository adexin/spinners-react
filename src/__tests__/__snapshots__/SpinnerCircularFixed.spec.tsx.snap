// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerCircularFixed matches the snapshot 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(255, 255, 255); overflow: visible; width: 100px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="rgba(0,0,0,0.44)"
      stroke-width="2"
    />
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="currentColor"
      stroke-dasharray="40, 134"
      stroke-dashoffset="325"
      stroke-linecap="round"
      stroke-width="2"
      style="animation: spinners-react-circular-fixed 14s linear infinite;"
    />
  </svg>
</div>
`;

exports[`SpinnerCircularFixed matches the snapshot with default props 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(56, 173, 72); overflow: visible; width: 50px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="rgba(0,0,0,0.44)"
      stroke-width="4"
    />
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="currentColor"
      stroke-dasharray="40, 134"
      stroke-dashoffset="325"
      stroke-linecap="round"
      stroke-width="4"
      style="animation: spinners-react-circular-fixed 1.4s linear infinite;"
    />
  </svg>
</div>
`;
